File: libs/typescript/tests/secrets.test.ts
import { test, expect, beforeEach } from "bun:test";
import { getSecret, _resetSecretsForTest } from "../src/secrets.ts";

beforeEach(() => {
  _resetSecretsForTest();
  delete process.env.SPADE_SECRETS;
});

test("returns injected secret value", () => {
  process.env.SPADE_SECRETS = JSON.stringify({ db: "postgres://user:pw@host/db" });
  expect(getSecret("db")).toBe("postgres://user:pw@host/db");
});

test("throws for a missing secret", () => {
  process.env.SPADE_SECRETS = JSON.stringify({ db: "x" });
  expect(() => getSecret("nope")).toThrow();
});

test("scrubs SPADE_SECRETS from the environment after load", () => {
  process.env.SPADE_SECRETS = JSON.stringify({ db: "x" });
  getSecret("db");
  expect(process.env.SPADE_SECRETS).toBeUndefined();
});

test("throws when no secrets were provided", () => {
  expect(() => getSecret("db")).toThrow();
});

File: libs/go/secrets_test.go
package spade

import (
	"os"
	"testing"
)

func resetSecrets() {
	secretsMu.Lock()
	secretsCache = nil
	secretsMu.Unlock()
}

func TestGetSecretReturnsValue(t *testing.T) {
	resetSecrets()
	t.Setenv("SPADE_SECRETS", `{"db":"postgres://user:pw@host/db"}`)
	v, err := GetSecret("db")
	if err != nil {
		t.Fatalf("GetSecret: %v", err)
	}
	if v != "postgres://user:pw@host/db" {
		t.Fatalf("got %q", v)
	}
}

func TestGetSecretMissing(t *testing.T) {
	resetSecrets()
	t.Setenv("SPADE_SECRETS", `{"db":"x"}`)
	if _, err := GetSecret("nope"); err == nil {
		t.Fatal("expected error for missing secret")
	}
}

func TestGetSecretScrubsEnv(t *testing.T) {
	resetSecrets()
	t.Setenv("SPADE_SECRETS", `{"db":"x"}`)
	if _, err := GetSecret("db"); err != nil {
		t.Fatalf("GetSecret: %v", err)
	}
	if _, ok := os.LookupEnv("SPADE_SECRETS"); ok {
		t.Fatal("SPADE_SECRETS should be scrubbed after load")
	}
}

func TestGetSecretNoEnv(t *testing.T) {
	resetSecrets()
	os.Unsetenv("SPADE_SECRETS")
	if _, err := GetSecret("db"); err == nil {
		t.Fatal("expected error when no secrets provided")
	}
}

func TestParseSecrets(t *testing.T) {
	if len(parseSecrets("")) != 0 {
		t.Fatal("empty blob should parse to empty map")
	}
	m := parseSecrets(`{"a":"1","b":"2"}`)
	if m["a"] != "1" || m["b"] != "2" {
		t.Fatalf("unexpected parse: %v", m)
	}
}
